README.md
# Flask EasyMode
[![Build Status](https://travis-ci.org/petermelias/flask-easymode.png?branch=master)](https://travis-ci.org/petermelias/flask-easymode) [![Coverage Status](https://coveralls.io/repos/petermelias/flask-easymode/badge.png?branch=master)](https://coveralls.io/r/petermelias/flask-easymode?branch=master) [![Downloads](https://pypip.in/d/flask-easymode/badge.png)](https://crate.io/packages/flask-easymode) [![Downloads](https://pypip.in/v/flask-easymode/badge.png)](https://crate.io/packages/flask-easymode)

## Motivation

This package was mainly created to consolidate and provide an API for common 
patterns that come up when using Flask for standard web applications.

Primarily it focuses on web API patterns, and view <--> model data handling 
patterns that generally do not change very often and can be very repetitive
at times.

## Design

This extension uses Blinker heavily to provide abstract plug points and assume
nothing about the application of any of the features provided by the API.
Generally speaking, you are expected to register Blinker recipient functions
in order to define your application specific behavior.

I highly recommend using ```@signal.connect``` and ```@signal.connect_via```
to register generic and increasingly specific handlers. This has the pleasant
effect of being efficient, lightweight, and decoupled.

I have made an effort to modularize and break up the various features so that
this package still follows the "opt-in" philosophy of Flask and its extensions.

For this reason, ```init_app()``` does almost nothing until you start enabling specific
features which will start touching config variables and the application context stack.

Much of the functionality is contained within mixins, helpers, and decorators which
must be imported after a feature has been enabled. Obviously there are certain
components that are usable directly and are not tied to a specific feature.

For example, the some of the mixins provided are mainly to assist with common
model class operations that come up on almost every project. These mixins
can just be plainly imported and used separately from even Flask itself.

That said, do not try and use decorators and/or helpers that depend on 
optionally-enabled features until you have enabled them. In some cases 
they will simply ignore you, in other cases they will throw errors.


## Features

* View Function Dependency-Injection of Models (by far the coolest feature)
* Fully Automatic XHR API Mode (error handling, data serialization, and more)
* Totally Abstract CRUD Mixins (examples using SQLAlchemy in Wiki)
* Useful View Helper functions such as ```redirect_self()``` and ```redirect_next()```

## Examples

### View Function DI

This example uses SQLAlchemy to look up our hypothetical user. Understand that the point
of using Signals here is that you can attach any lookup system you want. Totally abstract.

```python

# setup app
em = EasyMode()
em.init_app(app)
em.enable_injection(app)
em.add_injectable(User) # DO NOT FORGET TO DO THIS

# Sadly, EasyMode cannot magically know where you like to keep your injectable models
# If you have a class you want ID'd by an alternate name, do this:

em.add_injectable(User, alt='Uzer')

# Keep in mind if you load the same class many times with multiple 
# alternate names, they will all trigger loading of that same class
# if present during the actual injection phase. In some unusual cases
# this is actually desireable behavior.

# inherit user from Injectable mixin
class User(object, Injectable):
	
	def __init__(self, name):
		self.name = name

# subscribe database lookup to injection signal
@object_injected.connect_via(User)
def lookup_user(cls, conditions, **kwargs)
	query = session.query(cls)
	for attr, value in conditions:
		query.filter(getattr(cls, attr)==value)
	return query.first()
	# or redis
	# or cassandra
	# or a floppy disk

# setup routes
@app.route('/user/edit/<user_name>')
@inject('user')
def user_edit():
	return g.user.name # user is automagically assigned to g object (unless you turn that off)

# make a request
with app.test_client() as c:
	print c.get('/user/edit/petermelias') # petermelias
```

Super cool. Imagine all the code you can delete now because you don't have to lookup models
from your ```view_args```.

For anybody who is wondering, ```request.form``` is also checked for dependencies, so you can post params
as well and they will get picked up and injected right alongside url params. Merge behavior of course.

...and as of recently ```request.get_json()``` is ALSO checked for incoming parameters and merged as well. Trifecta.

If you don't like the auto-g-assignment, do this instead:

```python
@app.route('/user/edit/<user_name>')
@inject('user', as_args=True)
def user_edit(user):
	return user.name # injects as parameter instead of global, sexy I know.
```

Just when you thought it was over... you can specify unlimited conditions.

```python
@app.route('/users/list/<user_name>/<user_age>/<user_eye_color>')
@inject('user')
def users_list():
	print users # [User, User, User] all matching the parameters assuming your lookup function actually does that.

```

You can use arbitrarily long class names, the only rule is that the route params must be separated
by underscores between every word.

```python
class PharoahOfEgypt(object, Injectable): pass

@app.route('/pharoah/worship/<pharoah_of_egypt_name>')
@inject('pharoah_of_egypt')
def worship_pharoah():
	p = g.pharoah_of_egypt
```

### Full Auto XHR

This is really useful when you're communicating with a JavaScript frontend.

```python
# app
em = EasyMode()
em.init_app(app)
em.enable_xhr(app)

@app.route('/some-xhr-endpoint')
@xhr_api()
def xhr_endpoint():
	g.xhr.data['some-data'] = 'that I want to send to the client'

	flash('A super important message')
	flash('A huge problem', 'error')

	# you're not blind, there is no return statement necessary in this mode

with app.app_context() as c:
	r = c.get('/some-xhr-endpoint', headers=[('X-Requested-With', 'XMLHttpRequest')])
	print r.data

	'''
	{
		"data": {
			"some-data": "that I want to send to the client"
		},
		"messages": [
			["message", "A super important messsage"],
			["error", "A huge problem"]
		]
	}
	'''
```

There are many options and behavior notes for the XHR API but they are for specific
use cases that will be documented in the Wiki.

### Abstract CRUD Interface

Mixins and signals for attaching CRUD handlers. Also includes the Injectable mixin
and a bundled one if you want the super bonus 5-pack. See the CRUDI mixin.

```python

class User(CRUD): pass

# or if you want Injectable too
# class User(CRUDI): pass

# generic handling function for ALL models
@object_created.connect
def model_created(cls, o, **kwargs):
	data = kwargs.get('data')
	if data:
		o.update(**data) # update function comes from CRUD

# user specific handling function
@object_created.connect_via(User)
def user_created(cls, user, **kwargs):
	user.initialize_account()
	user.send_welcome_email()
	facebook.report_signup_to_nsa(user)

# generic model reading function using SQLAlchemy
@object_read.connect
def read_model(model, _many=False, **kwargs):
	r = None
	query = session.query(model)
	filters = kwargs.get('f')

	[query.filter(f) for f in filters if filters]

	if _many:
		r = query.all()
	else:
		r = query.first()
	return r

@object_deleted.connect
def delete_object(o, **kwargs):
	session.delete(o) # owned

# and in our view...

@app.route('/user/lookup/<user_name>')
@inject('user', as_args=True)
def lookup_user(user):
	user.update(**request.form.to_dict()) # because who needs validation anyway?
	user.delete() # or this user for that matter (comes from CRUD)

	some_other_user = User.read(filters=[User.name=='Mike', User.age>=12]) # read comes from CRUD
	user_list = User.read_many(filters=[User.age>=21]) # as does read_many (shortcuts _many=True)
 
	return jsonify(**some_other_user.as_dict) # as_dict and as_json come from CRUD

	# you can either attach your own as_dict / as_json functions or use the defaults
	# in future versions.
	# bear in mind that if you use the built in ones, you will need to define a 
	# _updateable and _readable property on your CRUD classes. This will not be
	# necessary in the next release.

```

As you can see this is both a fun, concise and robust way to set up data handling since it requires
the bare minimum amount of code to build an efficient framework for loading and modifying data.

In otherwords, no ActiveRecord bullshit here.

Also, keep in mind that ALL signalling uses ```**kwargs.``` You can make your handlers as simple and/or
as complex as needed. This library tries to assume the minimum level of convenience for the
consumer and uses private kwarg keys internally to avoid polluting your public API-space.

Fork, extend, merge, repeat!

## Tests

```python
nosetests
```

CHANGELOG.md
## Changelog
current version: v0.0.17

### v0.0.17
* Bugfix

### v0.0.16
* Bugfix

### v0.0.15
* Bugfix

### v0.0.14
* Minor change on ``` add_injectables ```

### v0.0.13
* Changed the ``` default ``` parameter to ``` nomatch ``` because "default" is highly ambiguous and I have no idea why it was that way. Current options are ``` [skip, load] ```

### v0.0.12
* Added the ``` lists='denote' ``` option to the ``` @inject ``` decorator. This default setting appends the class name of an injected result with the suffix '_list' to make it clear that a collection result is expected.

### v0.0.11
* Removed the ``` as_dict ``` and ``` as_json ``` interfaces from the ``` Read ``` model. Decided that was an overeach of responsibility for this extension. Recommend implementing in application space using something like ``` jsonpickle ``` library to create application-specific mixins for this behavior. Much cleaner, and more appropriately scoped.
* BREAKING: ``` delete ``` method on the ``` Delete ``` model is now a class method to match the others. The original motivation for making it an instance method was convenience. I have since decided that was a silly motivation and that for maximum ease of extensibility, class method (like the others) is the way to go. Consistency is also something I always want to maintain across my APIs and that was a factor in this decision. Old way: ``` obj.delete() ``` new way ``` ObjClass.delete(obj) ```
* You can now specify the sources you want to read from for injection ['json', 'params', 'form', 'query_string'] Use like this: ``` em.enable_injection(app, scan=('json', 'form')) ``` This would ignore the query string and the url parameters and only check POST and JSON data.
* Re-worked the internal configuration management model. Was driving me a little nuts. Do not read or write extension config variables on the ``` Flask.config ``` object. They are prefixed and one-way written from the extension. Use the ``` EasyMode.enable*() ``` functions to do stuff, the configuration values are really only for internal use by the extension.

### v0.0.10
* random fixes

### v0.0.9
* random fixes

### v0.0.8

* App context no longer used, pass application reference directly to enable*() functions.
* request.get_json() parsed for injection parameters.
* Alternate class names can be specified when adding new classes to the injector (this can happen as many times as you want)
* The Update Mixin has been modified to use a classmethod like the others and now operates on the instance as an argument.
  this changes the API from ``` o.update(**data) ``` to ``` ClassName.update(o, **data) ```. The reasoning behind this was to
  ensure that the consuming signal functions would have the same degree of flexibility in decision making as the other hook
  points for the CRUD system.
* Removed the default implementations for hook points like update(). It is not mission of the this library to make assumptions
  at that level. Even the default as_dict() implementation may be turned into a signal in later versions, and I may separate
  out the default impls into a module that can be activated or overridden at will. I would do this to provide the OOB experience
  for people in a hurry and also to allow for maximum choice for those looking to customize.

flask_easymode/tests/__init__.py
# -*- coding: utf-8 -*-


from flask import Flask, flash, g

from flask_easymode.decorators import xhr_api, inject
from flask_easymode.exceptions import XHRError
from flask_easymode.mixins.model import CRUD, CRUDI, object_injected


def create_app():
    app = Flask('easymode')

    app.config['TESTING'] = True
    app.config['SECRET_KEY'] = '123454fdsafdsfdsfdsfds'

    @app.route('/')
    def index():
        return 'I am the index page.'

    @app.route('/xhr')
    @xhr_api()
    def xhr_endpoint():
        try:
            g.xhr.data['test'] = 'monkey'
            g.xhr.data['walrus'] = 'punch'
        except AttributeError:
            pass

        flash('This is a test message.')
        flash('This is a warning.', 'warning')
        flash('This is an error', 'error')
        flash('This is just some info', 'info')
        flash('This is another warning', 'warning')

    @app.route('/xhr-failure')
    @xhr_api()
    def xhr_failure():
        try:
            g.xhr.data['test'] = 'monkey'
        except AttributeError:
            pass

        raise XHRError('Disaster everywhere.')

    @app.route('/xhr-failure-with-code')
    @xhr_api()
    def xhr_failure_with_code():
        try:
            g.xhr.data['test'] = 'monkey'
        except AttributeError:
            pass

        raise XHRError('Disaster befalls the city', status_code=500)

    @app.route('/xhr-that-returns-something')
    @xhr_api()
    def xhr_that_returns_something():
        try:
            g.xhr.data['test'] = 'monkey'
        except AttributeError:
            pass

        return 'Here is some string that would never be returned if the XHR API were active.'

    @app.route('/xhr-that-allows-regular-http')
    @xhr_api(allow_http=True)
    def xhr_that_allows_regular_http():
        try:
            g.xhr.data['test'] = 'monkey'
        except AttributeError:
            pass

        flash('A message in a bottle.')

        return 'Here is some regular return stuff'

    @app.route('/inject/<injectable_class_slug_name>')
    @inject('injectable_class')
    def inject_test_class_slug_name():
        return 'I have been injected with %s' % g.injectable_class.slug_name

    @app.route('/inject-as-arg/<injectable_class_slug_name>')
    @inject('injectable_class', as_args=True)
    def inject_test_class_args(injectable_class):
        return 'I have been injected with %s' % injectable_class.slug_name

    @app.route('/inject-non-injectable/<non_injectable_class_slug_name>')
    @inject('non_injectable_class')
    def inject_the_noninjectable():
        return 'This will never happen because there will be an exception :('

    @app.route('/inject-skip-by-default', defaults={'injectable_class_slug_name': None})
    @app.route('/inject-skip-by-default/<injectable_class_slug_name>')
    @inject('injectable_class', as_args=True)
    def inject_skip_by_default(injectable_class):
        return injectable_class.slug_name

    @app.route('/inject-list-denoting/<injectable_class_category_name>')
    @inject('injectable_class', default='skip', lists='denote', as_args=True)
    def inject_list_denoting(injectable_class_list):
        return str(injectable_class_list)

    return app


class InjectableClass(CRUDI):

    def __init__(self):
        self.slug_name = 'joe-slug'
        self.category_name = 'apples'


class NonInjectableClass(CRUD):
    pass


@object_injected.connect_via(InjectableClass)
def injectable_injected(cls, conditions, **kwargs):
    for c in conditions:
        k, v = c
        if k == 'slug_name' and v == 'joe-slug':
            return InjectableClass()

        if k == 'category_name' and v == 'apples':
            return [InjectableClass(), InjectableClass(), InjectableClass()]


flask_easymode/tests/test_decorators.py
# -*- coding: utf-8 -*-

from nose.tools.trivial import (
    assert_equals, assert_in, assert_raises)

from simplejson import loads

from flask_easymode import EasyMode
from flask_easymode.tests import (
    create_app, InjectableClass, NonInjectableClass)


def _setup():
    app = create_app()
    em = EasyMode()
    em.init_app(app)
    return (em, app)


def test_xhr_api():
    em, app = _setup()
    em.enable_xhr(app)

    with app.test_client() as c:
        r = c.get('/')
        assert_in('I am the index page', r.data)

        r = c.get('/xhr', headers=[('X-Requested-With', 'XMLHttpRequest')])

        py_data = loads(r.data)
        assert_in('data', py_data.keys())
        assert_in('messages', py_data.keys())

        assert_equals(len(py_data['data']), 2)
        assert_equals(len(py_data['messages']), 5)

        r = c.get('/xhr-failure')
        py_data = loads(r.data)
        assert_in('error', py_data.keys())

        r = c.get('/xhr-failure', headers=[
                  ('X-Requested-With', 'XMLHttpRequest')])
        py_data = loads(r.data)
        assert_in('error', py_data.keys())

        r = c.get('/xhr-failure-with-code', headers=[
                  ('X-Requested-With', 'XMLHttpRequest')])
        assert_equals(r.status_code, 500)
        py_data = loads(r.data)
        assert_in('error', py_data.keys())
        assert_in('code', py_data.keys())

        r = c.get('/xhr-that-allows-regular-http')
        py_data = loads(r.data)
        assert_in('test', py_data['data'].keys())
        assert_in('A message in a bottle.', py_data['messages'][0])


def test_xhr_api_off():
    em, app = _setup()

    with app.test_client() as c:
        r = c.get('/')
        assert_in('I am the index page.', r.data)

        r = c.get('/xhr')
        assert_in('403', r.data)

        r = c.get('/xhr', headers=[('X-Requested-With', 'XMLHttpRequest')])
        assert_in('403', r.data)

        r = c.get('/xhr-that-returns-something')
        assert_in('some string', r.data)


def test_inject():
    em, app = _setup()
    em.enable_xhr(app)
    em.enable_injection(app)
    em.add_injectable(InjectableClass)

    with app.test_client() as c:
        r = c.get('/inject/joe-slug')
        assert_in('joe-slug', r.data)

        r = c.get('/inject-as-arg/joe-slug')
        assert_in('joe-slug', r.data)

        with assert_raises(RuntimeError):
            r = c.get('/inject-non-injectable/here-comes-an-error')

        em.add_injectable(NonInjectableClass)

        with assert_raises(RuntimeError):
            r = c.get('/inject-non-injectable/still-going-to-fail')

        with assert_raises(AttributeError):
            r = c.get('/inject-skip-by-default')

        r = c.get('/inject-list-denoting/apples')
        assert_equals(r.status_code, 200)


def test_inject_off():
    em, app = _setup()

    with app.test_client() as c:
        with assert_raises(Exception):
            r = c.get('/inject/only-bad-can-come-of-this')


flask_easymode/tests/test_mixins/test_model.py
# -*- coding: utf-8 -*-

from nose.tools.trivial import (
    assert_equals, assert_true, assert_is_instance, assert_in)

from flask_easymode.mixins.model import (CRUD, object_created, object_read,
                                         object_updated, object_deleted)


class CRUDMan(CRUD):

    _readable = ('name', 'catch_phrase')
    _updateable = ('name')

    def __init__(self):
        self.name = 'CRUD Man'
        self.catch_phrase = 'I am CRUD Man. Fear my power.'


def test_create():

    @object_created.connect
    def create_fxn(cls, o, **kwargs):
        o.touched_by_signal = True

    o = CRUDMan.create()

    assert_true(o.touched_by_signal)
    assert_is_instance(o, CRUDMan)


def test_read():

    @object_read.connect
    def reading_fxn(cls, **kwargs):
        assert_equals(cls, CRUDMan)
        assert_in('name', kwargs.keys())

        if kwargs.get('_many'):
            return 'Many things'
        return 'Something'

    r = CRUDMan.read(name='CRUD Man')
    assert_equals(r, 'Something')

    r = CRUDMan.read_many(name='CRUD Man')
    assert_equals(r, 'Many things')


def test_update():

    data_dict = {'name': 'CRUD Woman', 'catch_phrase': 'Cant change this'}

    @object_updated.connect
    def updating_fxn(cls, o, **kwargs):
        assert_equals(o.name, 'CRUD Man')
        assert_equals(o.catch_phrase, 'I am CRUD Man. Fear my power.')
        return o

    o = CRUDMan.create()

    r = CRUDMan.update(o, **data_dict)
    assert_equals(r.name, 'CRUD Man')


def test_delete():

    o = CRUDMan.create()
    o.name = 'Dying CRUD Man'

    @object_deleted.connect
    def delete_fxn(cls, o, **kwargs):
        assert_equals(o.name, 'Dying CRUD Man')

    CRUDMan.delete(o)


flask_easymode/__init__.py
# -*- coding: utf-8 -*-

import re

from flask_easymode.exceptions import XHRError, handle_xhr_error

from mrpython.fxn import enum

_CONFIG_PREFIX = 'EM'


def _options():
    return enum(
        XHR_API_ENABLE=False,
        XHR_API_ALLOW_HTTP=False,
        DI_ENABLE=False,
        DI_SCAN=()
    )


class EasyMode(object):

    _injectables = {}
    _app_configs = {}

    def __init__(self, app=None):
        if app is not None:
            self.init_app(app)

    @classmethod
    def _set_config_option(cls, option, value, app):
        config = cls._lookup_config(app)
        if hasattr(config, option):
            setattr(config, option, value)
            cls._sync_config(config, app)

    @classmethod
    def _lookup_config(cls, app):
        try:
            return cls._app_configs[app]
        except KeyError:
            cls._app_configs[app] = _options()
            return cls._app_configs[app]

    @classmethod
    def _sync_config(cls, config, app):
        prefixed = {_CONFIG_PREFIX + '_' + k:
                    v for k, v in config.items.iteritems()}
        app.config.update(prefixed)

    @classmethod
    def _get_config_option(cls, option, app):
        config = cls._lookup_config(app)
        return getattr(config, option)

    @classmethod
    def config(cls, app):
        return cls._lookup_config(app)

    def init_app(self, app):
        EasyMode._sync_config(EasyMode._lookup_config(app), app)

    # as dirty as this looks, it's the best way to do it without using the decorator
    # because there is no built in function in flask yet to do it.
    def enable_xhr(self, app):
        self._set_config_option('XHR_API_ENABLE', True, app)
        app.error_handler_spec.setdefault(None, {}).setdefault(None, []) \
            .append((XHRError, handle_xhr_error))

    def enable_injection(self, app, scan=('json', 'form', 'params', 'query_string')):
        self._set_config_option('DI_ENABLE', True, app)
        self._set_config_option('DI_SCAN', scan, app)

    def add_injectable(self, cls, alt=None):
        cls_name = alt or cls.__name__
        cls_name = re.sub(
            r'(?<=[a-z])[A-Z]|(?<!^)[A-Z](?=[a-z])', r"_\g<0>",
            cls_name).lower()
        EasyMode._injectables[cls_name] = cls

    def add_injectables(self, *classes):
        for cls in classes:
            c, alt = cls, None
            try:
                c, alt = cls
            except (TypeError, ValueError):
                pass
            self.add_injectable(c, alt)


flask_easymode/mixins/model.py
# -*- coding: utf-8 -*-

from blinker import Namespace

easy_signals = Namespace()

object_created = easy_signals.signal('object_created')
object_read = easy_signals.signal('object_read')
object_updated = easy_signals.signal('object_updated')
object_deleted = easy_signals.signal('object_deleted')
object_injected = easy_signals.signal('object_injected')


class Create(object):

    @classmethod
    def create(cls, **kwargs):
        o = cls()
        object_created.send(cls, o=o, **kwargs)
        return o


class Read(object):

    @classmethod
    def read(cls, **kwargs):
        r = object_read.send(cls, **kwargs)
        try:
            return r[0][1]
        except IndexError:
            pass

    @classmethod
    def read_many(cls, **kwargs):
        r = object_read.send(cls, _many=True, **kwargs)
        try:
            return r[0][1]
        except IndexError:
            pass


class Update(object):

    @classmethod
    def update(cls, o, **kwargs):
        r = object_updated.send(cls, o=o, **kwargs)
        try:
            return r[0][1]
        except IndexError:
            pass


class Delete(object):

    @classmethod
    def delete(cls, o, **kwargs):
        r = object_deleted.send(cls, o=o, **kwargs)
        try:
            return r[0][1]
        except IndexError:
            pass


class CRUD(Create, Read, Update, Delete):
    pass


class Injectable(object):

    @classmethod
    def load(cls, conditions, **kwargs):
        r = object_injected.send(cls, conditions=conditions, **kwargs)
        try:
            return r[0][1]
        except IndexError:
            pass


class CRUDI(CRUD, Injectable):
    pass


flask_easymode/decorators.py
# -*- coding: utf-8 -*-

from flask import (current_app, request,
                   g, abort, get_flashed_messages, jsonify)

from functools import wraps
from collections import namedtuple

from flask_easymode import EasyMode
from flask_easymode.exceptions import XHRError

xhr = namedtuple('xhr', ('data', 'messages', 'html'))


class XHR(object):

    def __init__(self):
        self.html = ''
        self.data = {}
        self.messages = []


def xhr_api(allow_http=None):
    def _decorator(f):
        @wraps(f)
        def _wrapper(*args, **kwargs):
            if not EasyMode.config(current_app).XHR_API_ENABLE:
                view_result = f(*args, **kwargs)
                if not view_result:
                    abort(403)
                return view_result

            a = allow_http or EasyMode.config(current_app).XHR_API_ALLOW_HTTP

            if not request.is_xhr and not a:
                raise XHRError(
                    'XHR endpoints must be called asynchronously.',
                    status_code=500)

            g.xhr = XHR()
            f(*args, **kwargs)
            g.xhr.messages.extend(get_flashed_messages(with_categories=True))

            return jsonify(data=g.xhr.data,
                           messages=g.xhr.messages,
                           html=g.xhr.html)
        return _wrapper
    return _decorator


def inject(*classes, **options):
    def _decorator(f):
        @wraps(f)
        def _wrapper(*args, **kwargs):
            if not EasyMode.config(current_app).DI_ENABLE:
                return f(*args, **kwargs)

            # most to least specific to eliminate ambiguity when dealing with
            # nested prefix class names
            classes_by_len = sorted(classes, key=len)
            classes_by_len.reverse()

            injections = {}
            options.setdefault('nomatch', 'skip')
            options.setdefault('lists', 'denote')

            def _param_to_cls_prop_pair(param):
                for cls_name in classes_by_len:
                    r = param.replace(cls_name, '')
                    if r != param:
                        return (cls_name, r[1:])

            def _extract_injections(kvps):
                for k, v in kvps.iteritems():
                    try:
                        cls_name, prop_name = _param_to_cls_prop_pair(k)
                    except TypeError:
                        continue

                    injections[cls_name]['params'].append(k)
                    injections[cls_name]['conditions'].append((prop_name, v))

            for cls_name in classes:
                try:
                    cls = EasyMode._injectables[cls_name]
                except KeyError:
                    raise RuntimeError(
                        'Class "%s" has not been added as injectable. \
                        Use EasyMode.add_injectable(cls_name).' % cls_name)

                injections.setdefault(
                    cls_name, {'class': cls, 'params': [], 'conditions': []})

            sources = {
                'json': request.get_json(silent=True) or {},
                'form': request.form.to_dict(),
                'query_string': request.args,
                'params': kwargs
            }
            scan = EasyMode.config(current_app).DI_SCAN

            [_extract_injections(sources[s]) for s in scan]

            def _load(cls, cnd=[]):
                try:
                    o = cls.load(cnd)
                except AttributeError:
                    raise RuntimeError(
                        'To use %s with dependency injection, the \
                        class must either define a load(cls, \
                        conditions, **kwargs) interface or just \
                        inherit from the provided mixin.'
                        % cls.__name__)
                return o

            for cls_name, i in injections.iteritems():
                if cls_name in classes:
                    o = None
                    nomatch = options.get('nomatch')
                    if not i['conditions'] and nomatch == 'load':
                        o = _load(i['class'])

                    if i['conditions']:
                        o = _load(i['class'], i['conditions'])

                        if not o and nomatch == 'error':
                            return abort(400)

                    if type(o) is list and options.get('lists') == 'denote':
                        cls_name = cls_name + '_list'

                    if options.get('as_args'):
                        kwargs[cls_name] = o
                    else:
                        setattr(g, cls_name, o)

                    [kwargs.pop(p, None) for p in i['params']]

            return f(*args, **kwargs)
        return _wrapper
    return _decorator
